File: algo-dsa-beginner/dynamic-array/python/dynamic-array.py
class DynamicArray:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.buff = [0] * capacity


    def get(self, i: int) -> int:
        if i >= 0 and i < self.size:
            return self.buff[i]
        else:
            raise Exception("Tried to access invalid index: " + str(i))


    def set(self, i: int, n: int) -> None:
        if i >= 0 and i < self.size:
            self.buff[i] = n
        else:
            raise Exception("Tried to access invalid index: " + str(i))


    def pushback(self, n: int) -> None:
        if self.size == self.capacity:
            self.resize()

        self.buff[self.size] = n
        self.size += 1


    def popback(self) -> int:
        if self.size == 0:
            raise Exception("Can't pop element from empty list")
        else:
            ret = self.buff[self.size - 1]
            self.size -= 1
            return ret


    def resize(self) -> None:
        self.capacity *= 2
        new_buff = [0] * self.capacity
        for i in range(self.size):
            new_buff[i] = self.buff[i]
        self.buff = new_buff


    def getSize(self) -> int:
        return self.size


    def getCapacity(self) -> int:
        return self.capacity

def print_array(arr):
    print('size: ' + str(arr.getSize()) + ', capacity: ' + str(arr.getCapacity()))
    print('[', end=' ')
    for i in range(arr.getSize()):
        print(arr.get(i), end=' ')
    print(']')

my_vector = DynamicArray(1)

print_array(my_vector)

my_vector.pushback(1)
print_array(my_vector)

my_vector.pushback(2)
print_array(my_vector)

my_vector.pushback(3)
print_array(my_vector)

my_vector.popback()
print_array(my_vector)

my_vector.set(0, -1)
print_array(my_vector)

File: algo-dsa-beginner/03_dynamic-array/cpp/dynamic-array.cpp
#include <print>

class DynamicArray {
    int capacity;
    int size;
    int* buff;

    void resize() {
        this->capacity *= 2;

        int* new_buff = new int[this->capacity];
        for (int i = 0; i < this->size; i++) {
            new_buff[i] = this->buff[i];
        }

        delete[] this->buff;
        this->buff = new_buff;
    }

public:

    DynamicArray(int capacity)
    : capacity(capacity), size(0), buff(nullptr)
    {
       buff = new int[capacity];
    }

    ~DynamicArray() { delete[] buff; }

    int get(int i) const {
        return this->buff[i];
    }

    void set(int i, int n) {
        this->buff[i] = n;
    }

    void pushback(int n) {
        if (this->size == this->capacity) {
            this->resize();
        }

        this->buff[this->size++] = n;
    }

    int popback() {
        return this->buff[--this->size];
    }

    int getSize() const {
        return this->size;
    }

    int getCapacity() const {
        return this->capacity;
    }
};

void printVec(const DynamicArray& arr) {
    std::print("vec size = {}, ", arr.getSize());
    std::println("vec capacity = {}", arr.getCapacity());

    std::print("[ ");
    for (int i = 0; i < arr.getSize(); i++) {
        std::print("{} ", arr.get(i));
    }
    std::println("]");
}

int main() {
    DynamicArray vec{1};

    printVec(vec);

    vec.pushback(0);
    printVec(vec);

    vec.pushback(1);
    printVec(vec);

    vec.pushback(2);
    printVec(vec);

    vec.pushback(3);
    printVec(vec);

    vec.popback();
    printVec(vec);

    vec.set(0, 5);
    printVec(vec);
}
